vision/lenet_mnist/lenet_mnist.jl

```diff
@@ -0,0 +1,177 @@
+## Classification of MNIST dataset 
+## with the convolutional neural network know as LeNet5.
+## This script also combines various
+## packages from the Julia ecosystem  with Flux.
+using Flux
+using Flux.Data: DataLoader
+using Flux.Optimise: Optimiser, WeightDecay
+using Flux: onehotbatch, onecold, logitcrossentropy
+using Statistics, Random
+using Parameters: @with_kw
+using Logging: with_logger, global_logger
+using TensorBoardLogger: TBLogger, tb_overwrite, set_step!, set_step_increment!
+import ProgressMeter
+import MLDatasets
+import DrWatson: savename, struct2dict
+import BSON
+using CUDAapi
+
+# LeNet5 "constructor". 
+# The model can be adapted to any image size
+# and number of output classes.
+function LeNet5(; imgsize=(28,28,1), nclasses=10) 
+    out_conv_size =  (imgsize[1]÷4 - 3, imgsize[2]÷4 - 3, 16)
+    
+    return Chain(
+            x -> reshape(x, imgsize..., :),
+            Conv((5, 5), imgsize[end]=>6, relu),
+            MaxPool((2,2)),
+            Conv((5, 5), 6=>16, relu),
+            MaxPool((2,2)),
+            x -> reshape(x, :, size(x, 4)),
+            Dense(prod(out_conv_size), 120, relu), 
+            Dense(120, 84, relu), 
+            Dense(84, nclasses)
+          )
+end
+
+function get_data(args)
+    xtrain, ytrain = MLDatasets.MNIST.traindata(Float32, dir=args.datapath)
+    xtest, ytest = MLDatasets.MNIST.testdata(Float32, dir=args.datapath)
+
+    # MLDatasets uses HWCN format, Flux works with WHCN 
+    xtrain = permutedims(reshape(xtrain, 28, 28, 1, :), (2, 1, 3, 4))
+    xtest = permutedims(reshape(xtest, 28, 28, 1, :), (2, 1, 3, 4))
+
+    ytrain, ytest = onehotbatch(ytrain, 0:9), onehotbatch(ytest, 0:9)
+
+    train_loader = DataLoader(xtrain, ytrain, batchsize=args.batchsize, shuffle=true)
+    test_loader = DataLoader(xtest, ytest,  batchsize=args.batchsize)
+    
+    return train_loader, test_loader
+end
+
+loss(ŷ, y) = logitcrossentropy(ŷ, y)
+
+function eval_loss_accuracy(loader, model, device)
+    l = 0f0
+    acc = 0
+    ntot = 0
+    for (x, y) in loader
+        x, y = x |> device, y |> device
+        ŷ = model(x)
+        l += loss(ŷ, y) * size(x)[end]        
+        acc += sum(onecold(ŷ |> cpu) .== onecold(y |> cpu))
+        ntot += size(x)[end]
+    end
+    return (loss = l/ntot |> round4, acc = acc/ntot*100 |> round4)
+end
+
+## utility functions
+
+nobs(loader) = sum(size(x)[end] for (x, y) in loader) # == size(loader.data[1])[end] but more generic
+
+num_params(model) = sum(length(p) for p in Flux.params(model))
```

Comment:
```suggestion
num_params(model) = sum(length.(Flux.params(model)))
```
It might be my preference...